### PR TITLE
FIPS Support for Linux x86 RHEL

### PR DESCRIPTION
**Pull request by WilburZjh**
Signed-off-by: Jinhang Zhang <Jinhang.Zhang@ibm.com>

### CODE REVIEW

File: jdk/make/CopyFiles.gmk [hunk 1]
```diff
@@ -679,3 +679,17 @@ endif
 ##########################################################################################
 
 -include $(CUSTOM_MAKE_DIR)/CopyFiles.gmk
+
+################################################################################
+# Copy the nss.fips.cfg only on x86 linux
+
+ifeq ($(OPENJDK_TARGET_OS)-$(OPENJDK_TARGET_CPU_ARCH), linux-x86)
+  NSS_FIPS_CFG_SRC := $(TOPDIR)/closed/src/java.base/share/conf/security/nss.fips.cfg
```
Comment (keithc-ca) [inline jdk/make/CopyFiles.gmk:687]: This path is wrong, it should be:
```
  NSS_FIPS_CFG_SRC := $(TOPDIR)/closed/adds/jdk/src/share/conf/security/nss.fips.cfg
```

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: jdk/src/share/classes/sun/security/ec/SunEC.java [hunk 2]
```diff
@@ -23,12 +23,20 @@
...
+ */
+
 package sun.security.ec;
 
 import java.util.*;
 import java.security.*;
 import sun.security.action.PutAllAction;
 
+import openj9.internal.security.FIPSConfigurator;
+
```
[the remark this answers is not in the document]
Comment (WilburZjh) [reply jdk/src/share/classes/sun/security/ec/SunEC.java:39]: There are no engines (Key Factory engine, Algorithm Parameter engine, Signature engines, Key Pair Generator engine, etc) listed in SunEC in JDK8 like JDK11/JDK17. I just removed the `import` statement, and keep this `SunEC.java` file not changing.

File: jdk/make/CopyFiles.gmk [hunk 3]
```diff
@@ -679,3 +679,17 @@ endif
...
+  NSS_FIPS_CFG_DST := $(CONF_DST_DIR)/security/nss.fips.cfg
```
[the remark this answers is not in the document]
Comment (WilburZjh) [reply jdk/make/CopyFiles.gmk]: Edited.

File: closed/adds/jdk/src/share/classes/openj9/internal/security/FIPSConfigurator.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    private static final Debug debug = Debug.getInstance("semerufips");
+
+    // FIPS mode enable check, only supported on Linux x64.
+    private static final boolean userEnabledFIPS;
+    private static final boolean isFIPSSupported;
+    private static final boolean shouldEnableFIPS;
+
+    static {
+        String[] props = AccessController.doPrivileged(
+                new PrivilegedAction<>() {
```
[the remark this answers is not in the document]
Comment (WilburZjh) [reply closed/adds/jdk/src/share/classes/openj9/internal/security/FIPSConfigurator.java]: Edited.

File: jdk/src/share/classes/java/security/SecureRandom.java [hunk 5]
```diff
@@ -191,7 +199,24 @@ public SecureRandom(byte seed[]) {
...
+        String prngService = null;
+
+        // If in FIPS mode, use the SecureRandom from the FIPS provider.
+        if (FIPSConfigurator.enableFIPS()) {
+            Provider p = Security.getProvider("SunPKCS11-NSS-FIPS");
+            prng = "PKCS11";
+            if (p == null) {
+                throw new RuntimeException("could not find SunPKCS11-NSS-FIPS provider for FIPS mode");
+            }
+            prngService = p.getService("SecureRandom", prng);
```
[the remark this answers is not in the document]
Comment (WilburZjh) [reply jdk/src/share/classes/java/security/SecureRandom.java]: Edited.

File: jdk/src/share/classes/java/security/SecureRandom.java [hunk 6]
```diff
@@ -191,7 +199,24 @@ public SecureRandom(byte seed[]) {
     }
 
     private void getDefaultPRNG(boolean setSeed, byte[] seed) {
-        String prng = getPrngAlgorithm();
+        String prng = null;
```
[the remark this answers is not in the document]
Comment (WilburZjh) [reply jdk/src/share/classes/java/security/SecureRandom.java:202]: Edited.

File: jdk/make/CopyFiles.gmk (hunk 1 above)
Comment (WilburZjh) [reply jdk/make/CopyFiles.gmk]: Edited.

File: jdk/src/share/classes/java/security/SecureRandom.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (keithc-ca) [reply jdk/src/share/classes/java/security/SecureRandom.java:202]: This has not been addressed.

[the remark this answers is not in the document]
Comment (WilburZjh) [reply jdk/src/share/classes/java/security/SecureRandom.java]: Fixed.

File: closed/make/CopyFiles.gmk [hunk 7]
```diff
@@ -30,3 +30,17 @@ $(JDK_OUTPUTDIR)/include/ibmjvmti.h: $(SRC_ROOT)/openj9/runtime/include/ibmjvmti
 	$(call install-file)
 
 COPY_FILES += $(JDK_OUTPUTDIR)/include/ibmjvmti.h
+
+################################################################################
```
[the remark this answers is not in the document]
Comment (WilburZjh) [reply closed/make/CopyFiles.gmk]: removed.

**Comment by WilburZjh** [comment]
@keithc-ca @pshipton Please let me know if there are any other changes need to be made.

**Review by keithc-ca — APPROVED** [review]
These changes look reasonable to me, but I'll let @pshipton move this out of draft mode assuming he agrees it's time to do so.

**Comment by alon-sh** [comment]
we are looking to have FIPS mode support for jdk8 for the next release - therefore we need to merge this PR to have Lan enable testing for jdk8 so we can get started on the testing aspect of the work. 

FYI @mstoodle

**Review by pshipton — APPROVED** [review]

**Comment by pshipton** [comment]
This doesn't actually work so I'm going to revert it and a new PR can be created.

**Comment by pshipton** [comment]
Some review comments are:
- in jdk8 the java.security file is in ${java.home}/lib/security (not conf)
- the java.security specifies the full package names of the providers, i.e.
```
security.provider.1=sun.security.provider.Sun
security.provider.2=sun.security.rsa.SunRsaSign
security.provider.3=sun.security.ec.SunEC
security.provider.4=com.sun.net.ssl.internal.ssl.Provider
security.provider.5=com.sun.crypto.provider.SunJCE
```

**Comment by WilburZjh** [comment]
Issues fixed in #601

**Pull request merged**